src/libmongoc/src/mongoc/mongoc-cluster-aws.c

```diff
@@ -332,26 +369,44 @@ expiration_to_mcd_timer (const char *expiration_str,
       bson_destroy (&date_doc);
    }
 
-   // get current time in milliseconds since Unix Epoch.
-   int64_t now_ms;
-   {
-      struct timeval now;
-      if (0 != bson_gettimeofday (&now)) {
-         AUTH_ERROR_AND_FAIL ("bson_gettimeofday returned failure. Unable to "
-                              "determine expiration.");
-      } else {
-         now_ms =
-            (1000 * (int64_t) now.tv_sec) + ((int64_t) now.tv_usec / 1000);
-      }
+   if (!expiration_ms_to_timer (expiration_ms, expiration_timer, error)) {
+      goto fail;
    }
-
-   *expiration_timer = mcd_timer_expire_after (mcd_milliseconds (
-      expiration_ms - now_ms - MONGOC_AWS_CREDENTIALS_EXPIRATION_WINDOW_MS));
    ret = true;
 fail:
    return ret;
 }
 
+// generate_AWS_ROLE_SESSION_NAME generates a 16 byte hex string (32 characters)
+// to use as the AWS_ROLE_SESSION_NAME.
+static char *
+generate_AWS_ROLE_SESSION_NAME (bson_error_t *error)
+{
+#define NUM_BYTES 16
+   bool ok = false;
+   uint8_t data[NUM_BYTES];
+   char *out = bson_malloc (NUM_BYTES * 2 + 1);
+   if (!_mongoc_rand_bytes (data, NUM_BYTES)) {
+      AUTH_ERROR_AND_FAIL (
+         "unable to generate random bytes for AWS_ROLE_SESSION_NAME");
+   }
+
+   size_t i;
+   for (i = 0; i < NUM_BYTES; i++) {
+      bson_snprintf (out + (2 * i), 3, "%02x", data[i]);
+   }
+   out[NUM_BYTES * 2] = '\0';
```

Comment:
Suggest extracting this pattern (currently duplicated by `tests/unified/util.c`) into a `bin_to_hex` function in `mongoc-util-private.h`. Pairing this suggestion with the one requesting addition of test coverage for `bin_to_hex`.